### PR TITLE
Profile VM + UI

### PR DESCRIPTION
**Pull request by achillet3**
# Objective
Implementation of the Profile Screen. The code structure was changed to combine Friends' Profile Pages and the user's Profile Page into one VM + UI combo. This limits the duplication of code since both are very similar. The code thus uses quite some bits from the implementation from Nikhen of the Friends Profile Pages #283. 

# Implementation
- When clicking on a Friend's name in the chat screen, the friend's profile page is created through a factory. The friend's userId is passed as an argument to the factory and is then used to fetch that particular user's public data to display. 
- When clicking on the Profile Icon on the Navigation Bar, the factory is called with no argument. The View Model then understands that the user information to be fetched is that of the current user. 

# Notes
 - Text font sizes and colours have been changed to better contrast the label and the information. 
 - Location was removed from the implementation as it does not make much sense in this context (the investment time output did not seem great). Users would still be able to include that in their bio.

### CODE REVIEW

File: app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+      MutableStateFlow(
+          ProfileUiState(
+              profilePicUrl = "",
+              firstName = "Jim",
+              username = "Smith",
+              bio = "I am Jim Smith and I love the Smiths (the band)."))
+
+  @Before
+  fun testSetup() {
+    every { mockProfileViewModel.getProfileDetails() } returns Unit
```
Comment (Nath-Assa) [inline app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt:77]: Suggestion: To ensure the UI functions correctly with the VM, it is better to avoid mocking the VM. See PR #354 for details.

File: app/src/main/java/com/github/se/eventradar/viewmodel/AuthenticationViewModel.kt [hunk 2]
```diff
@@ -24,7 +24,7 @@ import kotlinx.coroutines.launch
 import kotlinx.coroutines.runBlocking
 
 @HiltViewModel
-class LoginViewModel @Inject constructor(private val userRepository: IUserRepository) :
+open class LoginViewModel @Inject constructor(private val userRepository: IUserRepository) :
```
Comment (Nath-Assa) [inline app/src/main/java/com/github/se/eventradar/viewmodel/AuthenticationViewModel.kt:27]: I understand that making a class open allows it to be subclassed, but could you please clarify the specific reason for making LoginViewModel an open class in this context?

**Review by Nath-Assa — APPROVED** [review]
Great work! I added a few suggestions/nits/questions. I have tested the app and confirmed that it functions correctly.

File: app/src/main/java/com/github/se/eventradar/ui/navigation/NavGraph.kt [hunk 3]
```diff
@@ -17,11 +17,10 @@ import com.github.se.eventradar.ui.login.SignUpScreen
...
 import com.github.se.eventradar.ui.qrCode.QrCodeTicketUi
-import com.github.se.eventradar.ui.viewProfile.ViewFriendsProfileUi
-import com.github.se.eventradar.util.toast
+import com.github.se.eventradar.ui.viewProfile.ProfileUi
 import com.github.se.eventradar.viewmodel.ChatViewModel
 import com.github.se.eventradar.viewmodel.EventDetailsViewModel
-import com.github.se.eventradar.viewmodel.ViewFriendsProfileViewModel
+import com.github.se.eventradar.viewmodel.ProfileViewModel
 import com.github.se.eventradar.viewmodel.qrCode.ScanFriendQrViewModel
 import com.github.se.eventradar.viewmodel.qrCode.ScanTicketQrViewModel
```
Comment (Nath-Assa) [inline app/src/main/java/com/github/se/eventradar/ui/navigation/NavGraph.kt:26]: Line 32:  `val context = LocalContext.current` can be deleted

**Comment by Nath-Assa** [comment]
Coverage on the new code is 75.5%. This is acceptable as long as the overall coverage for our code remains above 80%.

**Comment by Nath-Assa** [comment]
Detected bug: The phone number is not properly saved in the database, with the country code missing. This issue originates from the sign-up process and is not related to this PR.

File: app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        // act: go back !
+        performClick()
+      }
+    }
+    // assert: the nav action has been called
+    verify { mockNavActions.goBack() }
+    confirmVerified(mockNavActions)
+  }
+
+  /*
```
[the remark this answers is not in the document]
Comment (achillet3) [reply app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt]: Tried to test some edge cases to no avail

File: app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+      profilePic { assertIsDisplayed() }
+      name { assertIsDisplayed() }
+      username { assertIsDisplayed() }
+      leftAlignedViewProfileColumn { assertIsDisplayed() }
+      bioLabelText { assertIsDisplayed() }
+      bioInfoText { assertIsDisplayed() }
+      phoneNumberBirthDateRow { assertIsDisplayed() }
+      phoneNumberColumn { assertIsDisplayed() }
+      phoneNumberLabelText { assertIsDisplayed() }
+      // phoneNumberInfoText { assertIsDisplayed() }
```
[the remark this answers is not in the document]
Comment (achillet3) [reply app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt]: I don't understand why but both these lines represent the second child of a component and are unreachable through tests. However, both are still covered through the other tests so they should be deleted.

File: app/src/main/java/com/github/se/eventradar/ui/navigation/NavGraph.kt [hunk 6]
```diff
@@ -80,9 +81,19 @@ fun NavGraph(
...
+                navArgument("friendUserId") { type = NavType.StringType },
+            )) {
+          val userId = it.arguments!!.getString("friendUserId")!!
+          val viewModel = ProfileViewModel.create(userId = userId)
+          ProfileUi(viewModel = viewModel, navigationActions = navActions, isPublicView = true)
+        }
     composable(Route.PROFILE) {
-      HomeScreen(navigationActions = navActions)
-      context.toast("Profile screen needs to be implemented")
+      val viewModel = ProfileViewModel.create()
```
[the remark this answers is not in the document]
Comment (achillet3) [reply app/src/main/java/com/github/se/eventradar/ui/navigation/NavGraph.kt:81]: Indeed, this allows the view model to know when to fetch the data from which user

File: app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+      goBackButton { assertIsDisplayed() }
+      bottomNav { assertIsDisplayed() }
+      centeredViewProfileColumn { assertIsDisplayed() }
+      profilePic { assertIsDisplayed() }
+      name { assertIsDisplayed() }
+      username { assertIsDisplayed() }
+      leftAlignedViewProfileColumn { assertIsDisplayed() }
+      bioLabelText { assertIsDisplayed() }
+      bioInfoText { assertIsDisplayed() }
+      // In public view, phone number and birth date should not be displayed
```
[the remark this answers is not in the document]
Comment (achillet3) [reply app/src/androidTest/java/com/github/se/eventradar/viewProfile/ProfileTest.kt:67]: I agree. With the current development, the separation between Private and Public profiles is not very valuable. However, this does make sense for future development and the phone number and birth date fields are the less legitimate public information (it can even be argued that you never want your phone number to be shared for example)

File: app/src/main/java/com/github/se/eventradar/ui/viewProfile/ProfileUi.kt [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+import com.github.se.eventradar.viewmodel.ProfileViewModel
+import kotlinx.coroutines.runBlocking
+
+@Composable
+fun ProfileUi(
+    isPublicView: Boolean,
+    viewModel: ProfileViewModel,
+    navigationActions: NavigationActions
+) {
+  viewModel.getProfileDetails()
```
[the remark this answers is not in the document]
Comment (achillet3) [reply app/src/main/java/com/github/se/eventradar/ui/viewProfile/ProfileUi.kt:51]: I don't understand why but this is the only way it is guaranteed that the current user information is correctly fetched from the database.

File: app/src/test/java/com/github/se/eventradar/ProfileViewModelUnitTest.kt [hunk 9]
```diff
@@ -58,58 +77,41 @@ class ViewFriendsProfileViewModelUnitTest {
...
+          return ProfileViewModel(userRepository, userId)
         }
       }
 
+  @OptIn(ExperimentalCoroutinesApi::class)
   @Before
   fun setUp() {
     userRepository = MockUserRepository()
     runBlocking { userRepository.updateUser(mockUser) }
     runBlocking { userRepository.addUser(mockUser) }
```
[the remark this answers is not in the document]
Comment (achillet3) [reply app/src/test/java/com/github/se/eventradar/ProfileViewModelUnitTest.kt]: Thanks for the suggestion and the indication of the implementation in the next comment. Has been changed.

File: app/src/main/java/com/github/se/eventradar/viewmodel/AuthenticationViewModel.kt (hunk 2 above)
Comment (achillet3) [reply app/src/main/java/com/github/se/eventradar/viewmodel/AuthenticationViewModel.kt]: Nice catch! I had been playing around to optimise the duplication of code between the LoginViewModel and the ProfileViewModel. It should be a private class you're right
